Cart

Yushgoel opened the pull request:
This pull request implements CART algorithm for decision Trees -  Decision Tree Regressor  and Decision Tree Classifier. 

There is also an example file on how to use both algorithms.

Yushgoel committed:
Added Decision Tree Classifier

CART implementation of Decision Tree Classifier, based on Gini Impurity or Entropy, as selected by the user.

Yushgoel committed:
Adding Integration For Fixed Data Grid in Predict And Evaluate

Yushgoel committed:
Adding Regression Trees

Yushgoel committed:
Added Comments for Regressor

Yushgoel committed:
Adding Changes

Yushgoel committed:
Adding Example script for CART

Yushgoel committed:
Fixing Sorting

Yushgoel committed:
Updating Dataset + Naming

aravind-murthy-001 commented on examples/trees/cart.go:
```diff
@@ -0,0 +1,61 @@
+	// Evaluate Accuracy on Test Data
+	fmt.Println(decTree.Evaluate(testData))
+
+	// Load House Price Data For Regression
```
Remove empty line between comment and code.

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,481 @@
+	// Calculate loss based on Criterion Specified by user
+	if criterion == "gini" {
+		origGini, upperNode.LeftLabel = giniImpurity(y, labels)
+	} else if criterion == "entropy" {
```
Define "entropy" and "gini" as constants and use that everywhere. Check line 15 to 20 of id3.go

aravind-murthy-001 commented on examples/trees/cart.go:
```diff
@@ -0,0 +1,61 @@
+	regressionPreds := regTree.Predict(testRegData)
+
+	fmt.Println("Boston House Price Predictions")
+	fmt.Println(regressionPreds)
```
You can also document a section on how it compares to scikit, both in terms of the functioning of algorithm and performance wise (both training time and prediction time)

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,481 @@
+	return left, lefty, right, righty
+}
+
+// Fit - Method visible to user to train tree
```
Documentation of the method should at the high level explain what the method does. Not clear from some of these statements.

aravind-murthy-001 replied:
Check how it's done in id3 and other files.

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,481 @@
+
+	}
+	// Return the node - contains all information regarding feature and threshold.
+	return upperNode
```
This method has grown long. And seems to have some repeatitive code. Can you modularize this further.
Like the part on 
bestGini = subGini
bestLeft = left
bestRight = right
bestLefty = lefty

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,481 @@
+	Feature    int64
+	LeftLabel  int64
+	RightLabel int64
+	Use_not    bool
```
Why not use "UseNot".
Also the variable name is not conveying it's use.

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,481 @@
+					var rightGini float64
+					var leftLabels int64
+					var rightLabels int64
+
```
All four variables can be declared in one line.

aravind-murthy-001 replied:
Similarly at other places.

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,481 @@
+	if tree.Left == nil {
+		returnString += spacing + "---> True" + "\n"
+		returnString += "  " + spacing + "PREDICT    "
+		returnString += strconv.FormatInt(tree.LeftLabel, 10) + "\n"
```
Can remove empty lines at the begining and end of if blocks.

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,481 @@
+			}
+		}
+		p := float64(numLabel) / float64(nInstances)
+		gini += p * (1 - p)
```
Computing the gini impurity given a frequency distribution
 1 - sigma_i (p_i ^ 2)
is being done by computeGini Index method in gini.go.
Can't it be reused. 
Also the name of the methods giniImpurity, entropy is misleading if it is doing more than just computing gini Impurity of a distribution.

Yushgoel committed:
Fixing Comments

Yushgoel committed:
Adding cart_test.go

Yushgoel committed:
Removing Clutter

Partial Modularization of best split method. Shorten method by declaring variables in same line as well.

Also removing redundant functions, and adding into cart_utils.

Yushgoel committed:
Changing name of Use_not

Yushgoel committed:
Renaming Impurity Functions

Yushgoel committed:
Adding Documentation

Comparision in performance and implementation with sklearn.

aravind-murthy-001 commented on trees/cart_regressor.go:
```diff
@@ -0,0 +1,392 @@
+	classAttrs := X.AllClassAttributes()
+	// Only support 1 class Attribute
+	if len(classAttrs) != 1 {
+		panic(fmt.Sprintf("%d ClassAttributes (1 expected)", len(classAttrs)))
```
Never panic. All methods should return error.
Whenever there is error use fmt.Errorf("") and return.
When there is no error return nil.
Check id3.go for reference.

aravind-murthy-001 replied:
Everytime you call a method inside a method and the called method returns error, the caller should check for the error and return error if it receives an error. 
Example code in id3.go
serializer, err := base.CreateSerializedClassifierStub(filePath, metadata)
if err != nil {
	return  err
}

aravind-murthy-001 commented on trees/cart_regressor.go:
```diff
@@ -0,0 +1,392 @@
+	}
+	// Allocate return structure
+	_, rows := X.Size()
+	// labelVec := make([]float64, rows)
```
Remove commented out code.

aravind-murthy-001 commented on examples/trees/cart.go:
```diff
@@ -0,0 +1,76 @@
+
+		Sklearn:
+		Training Time for Titanic Dataset ≈ 8.8 µs
+		Prediction Time for Titanic Datset ≈ 7.87 µs
```
It is not clear with this if the difference in performance is constant, linear or exponential. 
Can you try with bigger dataset or like concatenate the same data set like 100 times and try.

aravind-murthy-001 commented on examples/trees/cart.go:
```diff
@@ -0,0 +1,76 @@
+		This implementation and sci-kit learn produce the exact same tree for the exact same dataset.
+		Predictions on the same test set also yield the exact same accuracy.
+
+		This implementation is optimized to prevent redundant iterations over the dataset, but it is not completely optimized. Also, sklearn makes use of numpy to access column easily, whereas here a complete iteration is required.
```
How can it be optimized further?

Yushgoel replied:
It's that it is easier and quicker to access a column of data in numpy, whereas here we have to iterate over the entire dataset to extract the feature.

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,415 @@
+
+// Calculate Gini Impurity of Target Labels
+func computeGiniImpurityAndModeLabel(y []int64, labels []int64) (float64, int64) {
+	nInstances := len(y)
```
Should return error if nInstances is 0. Else it will crash.

aravind-murthy-001 commented on examples/trees/cart.go:
```diff
@@ -0,0 +1,76 @@
+// Example of how to use CART trees for both Classification and Regression
```
Update package level details in this file.
https://github.com/Yushgoel/golearn/blob/master/trees/trees.go

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,415 @@
+}
+
+// Calculate Gini Impurity of Target Labels
+func computeGiniImpurityAndModeLabel(y []int64, labels []int64) (float64, int64) {
```
Not clear what y is. Prefer explicit naming of variables.

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,415 @@
+	gini := 0.0
+	maxLabelCount := 0
+	var maxLabel int64 = 0
+	for label := range labels {
```
If we need the value, we can ignore the index and access the value directly.
The variable name will also be intuitive. Can change it at other places to make it intuitive.

for _, label := range labels {
   for _, targetLabel := range y {
        if targetLabel == label {
            numLabel++
        }
   }
}

aravind-murthy-001 replied:
On further thought, if we have 10 labels and 1000 data points, this will be of the order of 10 * 1000.
You can instead create a map[string]int with label -> count by scanning the 1000 data points once and compute entropy on that.
Check if you can use getBaseEntropy or similar method from entropy.go that takes map[string]int

aravind-murthy-001 commented on trees/cart_regressor.go:
```diff
@@ -0,0 +1,392 @@
+		firstTime := true
+
+		var left, right [][]float64
+		var lefty, righty []float64
```
leftY rightY would be better.

aravind-murthy-001 commented on trees/cart_utils.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+// Isolate only unique values. This way, we can try only unique splits and not redundant ones.
+func findUnique(data []float64) []float64 {
```
Use map[string]bool to get unique values in list. More effiecient.
https://www.golangprograms.com/remove-duplicate-values-from-slice.html

aravind-murthy-001 replied:
Also reuse code anywhere possible either from 
https://github.com/Yushgoel/golearn/blob/master/utilities/utilities.go
or present in other places in code.

aravind-murthy-001 commented on trees/cart_classifier.go:
```diff
@@ -0,0 +1,415 @@
+	for label := range labels {
+		numLabel := 0
+		for target := range y {
+			if y[target] == labels[label] {
```
Similarly optimize here.

Yushgoel committed:
Removing Panics

Yushgoel committed:
Updating package level details

Yushgoel committed:
Optimizing Loss Calculation

Yushgoel committed:
catching nInstances == 0

Yushgoel committed:
Changing var name

Yushgoel committed:
Complexity Analysis for Algorithm

Yushgoel committed:
Updating Logistic.go
